### PR TITLE
dpcreconciler: set MAC and interface rename while interface is DOWN

### PR DESCRIPTION
**Pull request by rouming**
On attempt to set MAC address or rename interface which is UP host returns `Device or resource busy`. Patch fixes the error by changing the order: first set MAC and rename interface, then bring interface UP.

Patch fixes inability to use host interface, the problem was reproduced on RPI4 board. Presumably any device should be affected.

Fixes: 7a6981ba729e ("Add support for switch network instance with multiple ports")

PS. This PR is not tested. Need some comments from Milan, will do testing next week.

### CODE REVIEW

**Comment by OhmSpectator** [comment]
Did we see these errors in the test logs in the last few days? Or how have you noticed it?

File: pkg/pillar/dpcreconciler/linuxitems/adapter.go [hunk 1]
```diff
@@ -168,6 +161,15 @@ func (c *AdapterConfigurator) Create(ctx context.Context, item depgraph.Item) er
 		c.Log.Error(err)
 		return err
 	}
+	// Make sure that the interface is UP right after MAC address assignment
+	// and interface rename. In case of incorrect order operation fails
+	// with `Device or resource busy`.
+	if err := netlink.LinkSetUp(link); err != nil {
```
Comment (milan-zededa) [inline pkg/pillar/dpcreconciler/linuxitems/adapter.go:167]: I suggest to put this at the end (after `LinkSetUp(bridge)`) and with `kernLink` as the argument. In the past I saw `LinkSetMaster` changing the interface status to DOWN.

File: pkg/pillar/dpcreconciler/linuxitems/adapter.go [hunk 2]
```diff
@@ -131,13 +131,6 @@ func (c *AdapterConfigurator) Create(ctx context.Context, item depgraph.Item) er
 		// Managed by the wwan microservice, nothing to do here.
 		return nil
 	}
-	// Make sure that the interface is UP.
-	if err := netlink.LinkSetUp(link); err != nil {
```
Comment (milan-zededa) [inline pkg/pillar/dpcreconciler/linuxitems/adapter.go:135]: We need to make sure the interface is set UP even when it is not bridge by NIM, i.e. when `c.isAdapterBridgedByNIM(adapter)` returns false. Please add this into `if !c.isAdapterBridgedByNIM(adapter) { ... }`

**Review by milan-zededa — COMMENTED** [review]
Thanks for finding this, added some comments.

File: pkg/pillar/dpcreconciler/linuxitems/adapter.go (hunk 1 above)
Comment (rouming) [reply pkg/pillar/dpcreconciler/linuxitems/adapter.go]: Done.

File: pkg/pillar/dpcreconciler/linuxitems/adapter.go (hunk 2 above)
Comment (rouming) [reply pkg/pillar/dpcreconciler/linuxitems/adapter.go:135]: Done

File: pkg/pillar/dpcreconciler/linuxitems/adapter.go [hunk 3]
```diff
@@ -195,12 +195,21 @@ func (c *AdapterConfigurator) Create(ctx context.Context, item depgraph.Item) er
...
+	// otherwise host returns `Device or resource busy`.
 	if err := netlink.LinkSetUp(bridge); err != nil {
 		err = fmt.Errorf("netlink.LinkSetUp(%s) failed: %v",
 			adapter.IfName, err)
 		c.Log.Error(err)
 		return err
 	}
+	if err := netlink.LinkSetUp(kernLink); err != nil {
+		err = fmt.Errorf("netlink.LinkSetUp(%s) failed: %v",
+			adapter.IfName, err)
```
Comment (milan-zededa) [inline pkg/pillar/dpcreconciler/linuxitems/adapter.go:209]: Should be `kernIfname` in the error message.

Comment (rouming) [reply pkg/pillar/dpcreconciler/linuxitems/adapter.go]: Yep, thanks. Done.

**Review by milan-zededa — APPROVED** [review]
Starting eden tests. Let us know when you complete testing with the physical device.

**Comment by milan-zededa** [comment]
@eriknordmark Please do not make EVE release until we merge this.

**Comment by rouming** [comment]
Have just updated the "Radiatore" board (in Berlin-Lab), all good, keth0 appeared. 
cc: @rene

**Pull request merged**